common-libs/tiny-lds-ecdsa-secp256k1-2019/src/secp256k1signature.ts

```diff
@@ -0,0 +1,113 @@
+import { KEY_TYPE, VERIFICATION_KEY_TYPE, BufferLike, Signer, Verifier, BufferDataLike } from './common'
+import Secp256k1Key from './secp256k1key'
+
+const jsonld = require('jsonld')
+const jsigs = require('jsonld-signatures')
+
+const PROOF_SIGNATURE_KEY = 'jws'
+
+const stipParamsFromDidUrl = (did: string): string =>
+  did
+    // Strip out matrix params
+    .replace(/;[a-zA-Z0-9_.:%-]+=[a-zA-Z0-9_.:%-]*/g, '')
+    // Strip out query params
+    .replace(/([?][^#]*)?/g, '')
+
+type Secp256k1SignatureOptions = {
+  key: Secp256k1Key
+  date?: string
+  useNativeCanonize?: boolean
+}
+
+type ProofType = {
+  [PROOF_SIGNATURE_KEY]?: string
+  verificationMethod: string | { id: string }
+}
+type SignParams = { verifyData: BufferLike; proof: ProofType }
+type VerifyParams = { verifyData: BufferDataLike; proof: ProofType }
+
+export default class Secp256k1Signature extends jsigs.suites.LinkedDataSignature {
+  private readonly key: Secp256k1Key
+  private readonly verificationKeyType: string
+
+  private readonly signer: Signer
+  private readonly verifier: Verifier
+
+  verificationMethod: string
+
+  constructor(options: Secp256k1SignatureOptions) {
+    super({
+      type: KEY_TYPE,
+      LDKeyClass: Secp256k1Key,
+      date: options.date,
+      useNativeCanonize: options.useNativeCanonize,
+    })
+
+    this.verificationKeyType = VERIFICATION_KEY_TYPE
+
+    const publicKey = options.key.publicNode()
+    this.verificationMethod = publicKey.id
+    this.key = options.key
+    this.verifier = this.key.verifier()
+    this.signer = this.key.signer()
+  }
+
+  async sign({ verifyData, proof }: SignParams): Promise<ProofType> {
+    proof[PROOF_SIGNATURE_KEY] = await this.signer.sign({
+      data: verifyData,
+    })
+
+    return proof
+  }
+
+  async verifySignature({ verifyData, proof }: VerifyParams): Promise<boolean> {
+    const signature = proof[PROOF_SIGNATURE_KEY]
+    if (typeof signature !== 'string') {
+      throw new Error(`Expected proof[${PROOF_SIGNATURE_KEY}] to be string, got ${typeof signature}`)
```

Comment:
Introduced this check for typescript reasons (because according to type declarations, signature can be undefined).

It does not significantly change behavior of the code because `this.verifier.verify` calls to `signature.indexOf` unconditionally, and this call will throw an error if signature is undefined.